### DISCUSSION TITLE
Sliding at GB

### DISCUSSION THREAD

**Discussion by selarem** [idaholab/moose#17604] [Q&A Modules: Phase field] [answered]
Dear all,
In their work on Salt Rocks' deformation mechanism, colleagues have observed sliding at GB before the 
cracking starts. We are trying to model this feature under MOOSE. I have some development running (almost) fine with PF fracture coupled to  CP, but would like to know if there's a  way in moose to introduce this Sliding at GB without throwing all and restarting from scratch.
Thank you for any help
Saber

**Comment by jiangwen84** [comment, marked as answer]
How do you model your GB? sharp or diffuse interface? 

If it is sharp, I guess you can use cohesive zone method to model GB sliding.

**Reply by selarem** [reply]
I will be interested in both diffuse and sharp. At room temperature it is mainly sharp, but we expect different behaviour with increasing temperature.

**Reply by selarem** [reply]
Is there an already programmed meshModifier to split the mesh at the GB ? I am working with randomly generated grains.

**Reply by jiangwen84** [reply]
> Is there an already programmed meshModifier to split the mesh at the GB ? I am working with randomly generated grains.

Take a look at https://mooseframework.inl.gov/source/meshgenerators/BreakMeshByBlockGenerator.html

**Reply by selarem** [reply]
Back to moose after one year of Abaqus to discover that I could not split my mesh at the GB.
The grain number and properties are made at the MAterial block, how to tell the mesh breaker to create interfaces at GB ?

I have something like:

```
[Mesh]
[./fmg]
type = FileMeshGenerator
file=eprouvette.msh
[]
[]
[UserObjects]
[./prop_read]
type = ElementPropertyReadFile
prop_file_name = 'input_euler.txt'
nprop = 4
read_type = grain
ngrain = 200
rand_seed =7777
[../]
[]
```
and at the material block:
```
[./elasticity_tensor]
type = ComputeElasticityTensorCP
C_ijkl = '4.9592 2.670 2.670 4.9592 2.670 4.9592 3.111 3.111 3.111'
fill_method = symmetric9
read_prop_user_object = prop_read
[../]
```
Regards,

Saber

**Reply by GiudGiud** [reply]
Are you trying to create a sideset at the grain boundaries?
You can use a SideSetsBetweenSubdomainsGenerator in [Mesh] for that
Or partition the computation?

**Reply by selarem** [reply]
I want to use CZM at GB. So i need to generate
these side sets.
Regards

**Reply by GiudGiud** [reply]
Are the grains a separate subdomain? What do you see in the exodus output?
If so, then the mesh generator I pointed you to should work
https://mooseframework.inl.gov/source/meshgenerators/SideSetsBetweenSubdomainsGenerator.html

**Reply by selarem** [reply]
Lets take an example with 4 grains like in : ~/projects/moose/modules/tensor_mechanics/test/tests/elem_prop_read_user_object/prop_elem_read.i

is it possible in this case to introduce nodes to make interfaces for CZM to be considered at GB ?
regards,

**Reply by GiudGiud** [reply]
What are the grains in the example? I m not sure. 
This example has 4 elements and reads data per element. 

Is there a variable defined on these grains? A material property?
How do you know what the grain boundaries are?

**Reply by selarem** [reply]
That is my question from the beginning, how to tell the mesh modifier about the grain boundaries.

**Reply by GiudGiud** [reply]
Well the example you linked to is not about grains. 

I dived in the code a little bit, looks like the grains are determined by voronoi (nearest-neighbor) tesselation. So the short answer is that there currently is not an option for this, especially at the very beginning of the simulation since material properties/variables etc do not exist during the mesh generation. 
You could write your own mesh generator, or you could use this to do it after mesh generation:
https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html

The workflow would be:
- load the grain data
- copy the material property data loaded into an auxvariable using https://mooseframework.inl.gov/source/auxkernels/MaterialRealAux.html
- use the UO to change the subdomains based on the variable values

I have a PR that could simplify a little this workflow, where you could directly load the grain ID into a function, then use that instead of guessing based on material property values.
@loganharbour @dschwen if you have time to review it it would be appreciated. 
https://github.com/idaholab/moose/pull/19118

EDIT:
then once you have the blocks set, you can create the sidesets with the mesh generator I pointed to earlier

**Reply by selarem** [reply]
Thank for all these details. The problem of creating mesh and grains as blocks is solved by using neper. 
Now to move to the mechanics of the problem, I do not have any czm tests working. Are the tests just illustrative of the czm use ? 
regards

**Reply by GiudGiud** [reply]
Hello

What do you mean by they are not working? What error are you getting?

Guillaume

**Reply by GiudGiud** [reply]
@selarem any update on this?

**Comment by selarem** [comment]
czm_traction_separation_base.i
czm_patch_test_base.i
czm_multiple_dimension_base.i
czm_multiple_action_and_materials.i
All of them give error message. We need to modify them to run.

**Reply by GiudGiud** [reply]
They should run with an updated MOOSE?
What error message are you getting?

**Comment by selarem** [comment]
could not run it at that time. I will give it a try soon
